src/clients/BundleDataClient/BundleDataClient.ts

```diff
@@ -1043,37 +1059,106 @@ export class BundleDataClient {
+            // to the protocol.
+
+            // If fill exists in memory, then the only case in which we need to create a refund is if the
+            // the fill occurred in a previous bundle.
+            if (fill) {
+              if (!isSlowFill(fill) && fill.blockNumber < destinationChainBlockRange[0]) {
+                // If fill is in the current bundle then we can assume there is already a refund for it, so only
+                // include this pre fill if the fill is in an older bundle. If fill is after this current bundle, then
+                // we won't consider it, following the previous treatment of fills after the bundle block range.
+                validatedBundleV3Fills.push({
+                  ...fill,
+                  quoteTimestamp: deposit.quoteTimestamp,
+                });
+              }
+              return;
+            }
+
+            // If fill does not exist in memory but there is a slow fill request in memory, then we need to issue a
+            // slow fill leaf for the deposit. We can assume there was no fill preceding the slow fill request because
+            // slow fill requests cannot follow fills. If there were a fill following this request, we would have
+            // entered the above case. Again as with pre-fills, we should only consider slow fill requests that were
+            // in previous bundles.
+            if (slowFillRequest) {
+              if (_canCreateSlowFillLeaf(deposit) && slowFillRequest.blockNumber < destinationChainBlockRange[0]) {
+                validatedBundleSlowFills.push(deposit);
+              }
+              return;
+            }
+
+            // So at this point in the code, there is no fill or slow fill request in memory for this deposit.
+            // We need to check its fill status on-chain to figure out whether to issue a refund or a slow fill leaf.
+            // We can assume at this point that all fills or slow fill requests, if found, were in previous bundles
+            // because the spoke pool client lookback would have returned this entire bundle of events and stored
+            // them into the relay hash dictionary.
+            const fillStatus = await _getFillStatusForDeposit(deposit, destinationChainBlockRange[1]);
+
+            // If deposit was filled, then we need to issue a refund for it.
+            if (fillStatus === FillStatus.Filled) {
+              // We need to find the fill event to issue a refund to the right relayer and repayment chain,
+              // or msg.sender if relayer address is invalid for the repayment chain.
+              const prefill = (await findFillEvent(
+                destinationClient.spokePool,
+                deposit,
+                destinationClient.deploymentBlock,
+                destinationClient.latestBlockSearched
+              )) as unknown as FillWithBlock;
+              if (!isSlowFill(prefill)) {
```

Comment:
Maybe a log here as well. We'd definitely have to remove them once pre-fills become common, though.

Reply:
again i'm worried about adding a lot of noise. Imagine we enter here in a bundle, won't it get very noisy for that whole bundle?

Reply:
Yeah, it would have to be a debug log, and it would also have to be temporary so that once we actually do the rollout and pre-fills start happening, we can remove it. The idea was have it and make sure it never logged, but there are better ways to verify that this conditional never triggers now that I'm thinking about it.

Reply:
what kind of log are you proposing? or what are the better ways of verifying?

Reply:
Scratch the log. The better way to verify would be to run the this new dataworker in parallel with our current dataworker and verify that they produce the same bundles.